### PR TITLE
[Quant] Support lowering of channel shuffle in FX

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
## Description
Support lowering of channel shuffle in FX by adding its module and functional op to `is_copy_node` list in `torch/ao/quantization/fx/_lower_to_native_backend.py`

## Validation
UTs added to test
- correctness of quantized `ChannelShuffle` module.
- FX lowering of `ChannelShuffle` module and functional `channel_shuffle`.

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @SherlockNoMad @soumith @EikanWang @wenzhe-nrv

### CODE REVIEW

File: torch/nn/quantized/modules/channelshuffle.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import torch
+
+class ChannelShuffle(torch.nn.ChannelShuffle):
+    def __init__(self, groups: int):
```
Comment (vadimkantorov) [inline torch/nn/quantized/modules/channelshuffle.py:4]: is this constructor needed? since it just passes through the arguments

the base-derived constructor would be sufficient?

Comment (Xia-Weiwen) [reply torch/nn/quantized/modules/channelshuffle.py]: Thanks. They are removed.

**Comment by Xia-Weiwen** [comment]
@vadimkantorov Please review again. Thanks.

**Comment by vadimkantorov** [comment]
@Xia-Weiwen I'm not member of PyTorch, so a review from someone else is required :)

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen I'm not a member of PyTorch team, so a review from someone else is required :)

I see. Thank you anyway.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Could you please review this PR? Or could you ask someone else to review? Thanks!

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 2]
```diff
@@ -84,8 +84,10 @@ def is_default_node(node, modules):
         torch.nn.LayerNorm,
         torch.nn.Dropout,
         torch.nn.PReLU,
+        torch.nn.ChannelShuffle,
         torch.nn.BatchNorm2d,
         torch.nn.BatchNorm3d,
+        torch.nn.ChannelShuffle,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:90]: why do we have two channel shuffle here?

File: torch/nn/quantized/modules/channelshuffle.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import torch
+
+class ChannelShuffle(torch.nn.ChannelShuffle):
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/channelshuffle.py:3]: if the float version works for quantized inputs as well, I don't think we will need to define an extra quantized module

Comment (jerryzh168) [reply torch/nn/quantized/modules/channelshuffle.py:3]: it should be a part of this list in that case I think: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/_lower_to_native_backend.py#L122

File: torch/ao/quantization/backend_config/_common_operator_config_utils.py [hunk 4]
```diff
@@ -353,6 +353,7 @@ def _get_default_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPat
         torch.nn.LayerNorm,
         torch.nn.Dropout,
         torch.nn.PReLU,
+        torch.nn.ChannelShuffle,
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/_common_operator_config_utils.py:356]: should add this to _get_share_qprams_op_backend_config I think

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
channel shuffle should be categorized as copy node I think, so we don't need to add new quantized modules and configs and lowering need to change as well, please see inline comments for changes

**Comment by Xia-Weiwen** [comment]
> channel shuffle should be categorized as copy node I think, so we don't need to add new quantized modules and configs and lowering need to change as well, please see inline comments for changes

Thanks for reviewing. I have removed most of previous changes. Now I only add it to the `is_copy_node` list and add two unit tests. Please review again. Thanks.

File: torch/ao/quantization/backend_config/_common_operator_config_utils.py (hunk 4 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/backend_config/_common_operator_config_utils.py]: I have added channel shuffle to copy node list. Is this still needed?

File: torch/nn/quantized/modules/channelshuffle.py (hunk 3 above)
Comment (Xia-Weiwen) [reply torch/nn/quantized/modules/channelshuffle.py]: OK. I have removed them.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 2 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py]: Sorry for the typo. Now they are removed.

**Comment by yanbing-j** [comment]
Hi @jerryzh168 , could you please help review again? Thanks!

File: test/quantization/fx/test_quantize_fx.py [hunk 5]
```diff
@@ -4771,6 +4771,41 @@ def _test(prepare_fn, qconfig_dict):
...
+                return torch.nn.functional.channel_shuffle(x + x, 2) + x
+
+        x = torch.randn(4, 4, 4, 4)
+        for m in (M1().eval(), M2().eval(), M3().eval()):
+            m = prepare_fx(m, {"": default_qconfig}, example_inputs=(x,))
+            m_copy = copy.deepcopy(m)
+            m = convert_fx(m)
+            m_ref = convert_to_reference_fx(m_copy)
+            node_occurrence = {
+                ns.call_function(torch.quantize_per_tensor): 1,
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4799]: do we want to check that the op is quantized as well?

Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:4799]: in this case quantized just means we use the original op, but it would still be goo to explicitly test this

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 6]
```diff
@@ -111,6 +111,7 @@ def is_copy_node(node, modules):
         torch.flatten,
         torch.mean,
         operator.floordiv,
+        torch.channel_shuffle,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:114]: do we need to include torch.functional.channel_shuffle as well? and why this is not catched by the test?

**Comment by jerryzh168** [comment]
could you update the PR summary?

File: test/quantization/fx/test_quantize_fx.py (hunk 5 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:4803]: > do we want to check that the op is quantized as well?

OK. Now we check occurrences of the op itself. Do you mean that?

Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:4803]: Looks like `torch.channel_shuffle` and `torch.nn.functional.channel_shuffle` are actually the same thing. So, we only check occurrence of either `torch.channel_shuffle` or `torch.nn.functional.channel_shuffle`.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: Looks like `torch.channel_shuffle` and `torch.nn.functional.channel_shuffle` are actually the same thing:
```
>>> import torch
>>> print(torch.channel_shuffle)
<built-in method channel_shuffle of type object at 0x7f74ab048560>
>>> print(torch.nn.functional.channel_shuffle)
<built-in method channel_shuffle of type object at 0x7f74ab048560>
>>> print(torch.nn.ChannelShuffle)
<class 'torch.nn.modules.channelshuffle.ChannelShuffle'>
```

**Comment by Xia-Weiwen** [comment]
> could you update the PR summary?

Hi @jerryzh168 Sorry for the late reply. It's done. Please review again. Thanks.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: BTW, do you know where are `torch.channel_shuffle`/`torch.nn.functional.channel_shuffle` defined? I cannot find implementation in source code, only got this:
https://github.com/pytorch/pytorch/blob/09965957cd8ecc696852e73022892b3ad4475783/torch/nn/functional.py#L3576

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Please review. Thanks!

File: test/quantization/fx/test_quantize_fx.py [hunk 7]
```diff
@@ -5109,6 +5109,47 @@ def validate_qconfig(qconfig: QConfig, satisfies_constraints: bool):
...
+        for i in range(len(models)):
+            m = models[i]
+            m = prepare_fx(m, {"": default_qconfig}, example_inputs=(x,))
+            m_copy = copy.deepcopy(m)
+            m = convert_fx(m)
+            m_ref = convert_to_reference_fx(m_copy)
+            node_occurrence = {
+                ns.call_function(torch.quantize_per_tensor): 1,
+                ns.call_module(torch.nn.ChannelShuffle): 1 if i == 0 else 0,
+                ns.call_function(torch.channel_shuffle): 1 if i > 0 else 0,
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5141]: this is a bit confusing, could you move this to the loop, e.g.
```
expected_nodes = [ns.call_module(torch.nn.ChannelShuffle), ns.call_function(torch.channel_shuffle), ...]
for model, node in zip(models, expected_node)
   ...
```

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: this is where F.channel_shuffle is defined I think, it just says F.channel_shuffle is the same as torch.channel_shuffle

File: test/quantization/core/test_quantized_module.py [hunk 8]
```diff
@@ -1036,6 +1036,33 @@ def test_prelu(self):
         self.assertEqual(qy_ref, qy,
                          msg="PReLU module API failed")
 
+    def test_channel_shuffle(self):
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_module.py:1039]: do you want to add a operator test for torch.channel_shuffle/F.channel_shuffle?

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: maybe we can add a test saying `torch.nn.functional.channel_shuffle` is the same as `torch.channel_shuffle` so we only need to support lowering for `torch.channel_shuffle`?

File: test/quantization/core/test_quantized_module.py (hunk 8 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_module.py:1039]: I think we already have it: https://github.com/pytorch/pytorch/blob/dcac4dd58edefb6951a60266e53d8767dc9be002/test/quantization/core/test_quantized_op.py#L1298

File: test/quantization/fx/test_quantize_fx.py (hunk 7 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py]: Sure, it's done.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: Do you mean adding a line of comment or an assertion here? Or a separate test that only has one line of assertion?

Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: > Do you mean adding a line of comment or an assertion here? Or a separate test that only has one line of assertion?

Hi @jerryzh168 Please take a look

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Please review. Thanks!

File: test/quantization/fx/test_quantize_fx.py [hunk 9]
```diff
@@ -5263,6 +5263,50 @@ def forward(self, x):
...
+                return torch.nn.functional.channel_shuffle(x + x, 2) + x
+
+        x = torch.randn(4, 4, 4, 4)
+        models = [M1().eval(), M2().eval(), M3().eval()]
+        # torch.channel_shuffle is torch.nn.functional.channel_shuffle
+        expected_nodes = [
+            ns.call_module(torch.nn.ChannelShuffle),
+            ns.call_function(torch.channel_shuffle),
+            ns.call_function(torch.channel_shuffle)
+        ]
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5291]: nit: might be better to just write a list of tuple of (model, expected_node) so that it's clearer

File: test/quantization/fx/test_quantize_fx.py (hunk 5 above)
Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5299]: I didn't see the expected node appear here, should we add that?

**Review by jerryzh168 — APPROVED** [review]
looks good, please add expected channel shuffle op to `node_occurrence` as well

File: test/quantization/fx/test_quantize_fx.py (hunk 9 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py]: OK. It's done. Please take a look.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 I changed per your comments. All checks have passed. Do you have more comments? Thanks.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:114]: I think you can add a comment

File: test/quantization/fx/test_quantize_fx.py (hunk 5 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:5330]: > I didn't see the expected node appear here, should we add that?

Do you mean line 5329?

Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5330]: oh I see, sorry I missed that, LGTM

**Review by jerryzh168 — APPROVED** [review]
LG, please add a comment in `torch/ao/quantization/fx/_lower_to_native_backend.py` about why we only need `torch.channel_shuffle` and not `F.channel_shuffle`

**Comment by Xia-Weiwen** [comment]
> LG, please add a comment in `torch/ao/quantization/fx/_lower_to_native_backend.py` about why we only need `torch.channel_shuffle` and not `F.channel_shuffle`

Thanks. I will add that comment and land it after all CI checks pass.

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 6 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:116]: Added
